### PR TITLE
Python: Update as_dict/from_dict

### PR DESCRIPTION
**Pull request by lucasmcdonald3**
*Issue #, if available:*

*Description of changes:*

* Fix as_dict/from_dict for SDK clients (just pass the client through)
* Override Smithy-Python to write dict parameters correctly (snakecase, see https://peps.python.org/pep-0008/#function-and-variable-names)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java [hunk 1]
```diff
@@ -411,6 +435,94 @@ protected void writeFromDict(boolean isError) {
     writer.write("");
   }
 
+  /**
+   * Override Smithy-Python writeAsDict to handle shapes with {@link ReferenceTrait}s.
+   * Most of this is lifted directly from Smithy-Python; the changed components are
+   * highlighted.
```
Comment (ShubhamChaturvedi7) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java]: The whole thing is highlighted if you meant that the diff would be highlighted (In that case, it's better to remove the highlighted word).

**Review by ShubhamChaturvedi7 — COMMENTED** [review]
Is CI expected to fail?

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java (hunk 1 above)
Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/extensions/DafnyPythonLocalServiceStructureGenerator.java]: Oh no, by "highlighted" I meant that I was calling out specific areas.
But it is unclear. Let me clarify

**Comment by lucasmcdonald3** [comment]
CI doesn't need to pass for this PR, but it was easy to make it pass. I updated this branch with the latest python-reviewed branch, and now most CI is passing.

The final check will pass after https://github.com/smithy-lang/smithy-dafny/pull/541, which formats the file in this PR with the updated logic from this PR, so I'll merge #541 in after this PR goes in.

**Review by ShubhamChaturvedi7 — APPROVED** [review]
Checked with Lucas, prettier runs here https://github.com/smithy-lang/smithy-dafny/pull/541

**Pull request merged**